Les outputs des graphiques de ce projet ont été retirés car ils rendaient le dépôt trop volumineux.
Elles sont visibles dans le support de présentation PowerPoint associé.

<div style="background-color: RGB(0,150,250);" >
<h2 style="margin: auto; padding: 20px; color:#fff; ">1 - Importation des librairies et chargement des fichiers</h2>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">1.1 - Importations des librairies</h3>
</div>

In [30]:
#Importation des librairies

#Importation de la librairie Numpy
import numpy as np

#Importation de la librairie Pandas
import pandas as pd

#Importation de la libraire Plotly
import plotly.express as px

px.defaults.template='plotly_dark'
px.defaults.width = 1200
px.defaults.height = 450

#Importation de la librairie Matplotlib
import matplotlib.pyplot as plt

#Importation de la librairie Datetime
import datetime as dt

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">1.2 - Chargement des fichiers</h3>
</div>

In [31]:
#Importation du ficher customers.csv
customers = pd.read_csv('../Ressources/Data/customers.csv', sep=';', dtype={'client_id':str,
                                                                            'sex':str,
                                                                            'birth':int})

#Importation du ficher products.csv
products = pd.read_csv('../Ressources/Data/products.csv', sep=';', dtype={'id_prod':str,
                                                                          'price':float,
                                                                          'categ':str})

#Importation du ficher transactions.csv
transactions = pd.read_csv('../Ressources/Data/Transactions.csv', sep=';',
                           dtype={'id_prod':str,'date':str,'session_id':str,'client_id':str})

<div style="background-color: RGB(0,150,250);" >
<h2 style="margin: auto; padding: 20px; color:#fff; ">2 - Analyse exploratoire des fichiers</h2>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">2.1 - Fichier customers</h3>
</div>

In [32]:
#Afficher les dimensions du dataset
print("Le tableau comporte {} ligne(s)".format(customers.shape[0]))
print("Le tableau comporte {} colonne(s)".format(customers.shape[1]))

Le tableau comporte 8621 ligne(s)
Le tableau comporte 3 colonne(s)


In [33]:
#Afficher les détails du dataset
print(customers.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8621 entries, 0 to 8620
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  8621 non-null   object
 1   sex        8621 non-null   object
 2   birth      8621 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 202.2+ KB
None


In [34]:
#Afficher les 5 premières lignes
customers.head()

,client_id,sex,birth
0,c_4410,f,1967
1,c_7839,f,1975
2,c_1699,f,1984
3,c_5961,f,1962
4,c_5320,m,1943


In [35]:
#Afficher les détails de la colonne client_id
print(customers['client_id'].describe())

count     8621
unique    8621
top       c_84
freq         1
Name: client_id, dtype: object


In [36]:
#Afficher les valeurs uniques de la colonne sex
customers['sex'].value_counts()


sex
f    4490
m    4131
Name: count, dtype: int64

In [37]:
#Afficher le min et le max de la colonne birth
print("Année de naissance la plus ancienne : {}".format(customers['birth'].min()))
print("Année de naissance la plus récente : {}".format(customers['birth'].max()))

Année de naissance la plus ancienne : 1929
Année de naissance la plus récente : 2004


<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">2.2 - Fichier products</h3>
</div>

In [38]:
#Afficher les dimensions du dataset
print("Le tableau comporte {} ligne(s)".format(products.shape[0]))
print("Le tableau comporte {} colonne(s)".format(products.shape[1]))

Le tableau comporte 3286 ligne(s)
Le tableau comporte 3 colonne(s)


In [39]:
#Afficher les détails du dataset
print(products.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3286 entries, 0 to 3285
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id_prod  3286 non-null   object 
 1   price    3286 non-null   float64
 2   categ    3286 non-null   object 
dtypes: float64(1), object(2)
memory usage: 77.1+ KB
None


In [40]:
#Afficher les 5 premières lignes
products.head()

,id_prod,price,categ
0,0_1421,19.99,0
1,0_1368,5.13,0
2,0_731,17.99,0
3,1_587,4.99,1
4,0_1507,3.99,0


In [41]:
#Afficher les détails de la colonne id_prod
print(products['id_prod'].describe())

count       3286
unique      3286
top       0_1920
freq           1
Name: id_prod, dtype: object


In [42]:
#Afficher les détails de la colonne price
print(products['price'].describe())

count    3286.000000
mean       21.863597
std        29.849786
min         0.620000
25%         6.990000
50%        13.075000
75%        22.990000
max       300.000000
Name: price, dtype: float64


In [43]:
#Afficher les valeurs uniques de la colonne categ
print(products['categ'].value_counts())

categ
0    2308
1     739
2     239
Name: count, dtype: int64


<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">2.3 - Fichier transactions</h3>
</div>

In [44]:
#Afficher les dimensions du dataset
print("Le tableau comporte {} ligne(s)".format(transactions.shape[0]))
print("Le tableau comporte {} colonne(s)".format(transactions.shape[1]))

Le tableau comporte 1048575 ligne(s)
Le tableau comporte 4 colonne(s)


In [45]:
#Afficher les détails du dataset
print(transactions.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   id_prod     687534 non-null  object
 1   date        687534 non-null  object
 2   session_id  687534 non-null  object
 3   client_id   687534 non-null  object
dtypes: object(4)
memory usage: 32.0+ MB
None


In [46]:
#Afficher les 5 premières lignes
transactions.head()

,id_prod,date,session_id,client_id
0,0_1259,2021-03-01 00:01:07.843138,s_1,c_329
1,0_1390,2021-03-01 00:02:26.047414,s_2,c_664
2,0_1352,2021-03-01 00:02:38.311413,s_3,c_580
3,0_1458,2021-03-01 00:04:54.559692,s_4,c_7912
4,0_1358,2021-03-01 00:05:18.801198,s_5,c_2033


In [47]:
#Afficher les détails de la colonne id_prod
print(transactions['id_prod'].describe())

count     687534
unique      3265
top        1_369
freq        2340
Name: id_prod, dtype: object


In [48]:
#Afficher les détails de la colonne session_id
print(transactions['session_id'].describe())

count       687534
unique      345505
top       s_118668
freq            14
Name: session_id, dtype: object


In [49]:
#Afficher les détails de la colonne client_id
print(transactions['client_id'].describe())

count     687534
unique      8600
top       c_1609
freq       25586
Name: client_id, dtype: object


In [50]:
#Compter le nombre de lignes null dans le dataset
print("Nombre de lignes null dans le dataset : \n{}".format(transactions.isnull().sum()))

Nombre de lignes null dans le dataset : 
id_prod       361041
date          361041
session_id    361041
client_id     361041
dtype: int64


In [51]:
#Supprimer les lignes avec des valeurs null
transactions = transactions.dropna()
print("Le tableau comporte {} ligne(s)".format(transactions.shape[0]))

Le tableau comporte 687534 ligne(s)


<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">2.4 - Fusion des dataframes</h3>
</div>

In [52]:
#Création d'un dataframe df en fusionnant les 3 dataframes
df = pd.merge(transactions, products, on='id_prod')
df = pd.merge(df, customers, on='client_id')

In [53]:
#Afficher les 5 premières lignes
df.head()

,id_prod,date,session_id,client_id,price,categ,sex,birth
0,0_1259,2021-03-01 00:01:07.843138,s_1,c_329,11.99,0,f,1967
1,0_1390,2021-03-01 00:02:26.047414,s_2,c_664,19.37,0,m,1960
2,0_1352,2021-03-01 00:02:38.311413,s_3,c_580,4.50,0,m,1988
3,0_1458,2021-03-01 00:04:54.559692,s_4,c_7912,6.55,0,f,1989
4,0_1358,2021-03-01 00:05:18.801198,s_5,c_2033,16.49,0,f,1956


<div style="background-color: RGB(0,150,250);" >
<h2 style="margin: auto; padding: 20px; color:#fff; ">3 - Analyse des différents indicateurs de vente</h2>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1 - Évolution dans le temps du chiffre d'affaires</h3>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1.1 - Chiffre d'affaires avec la moyenne mobile</h4>
</div>

In [54]:
#Création de la colonne month
df['month'] = pd.to_datetime(df['date']).dt.to_period('M').dt.to_timestamp()
df.head()

,id_prod,date,session_id,client_id,price,categ,sex,birth,month
0,0_1259,2021-03-01 00:01:07.843138,s_1,c_329,11.99,0,f,1967,2021-03-01
1,0_1390,2021-03-01 00:02:26.047414,s_2,c_664,19.37,0,m,1960,2021-03-01
2,0_1352,2021-03-01 00:02:38.311413,s_3,c_580,4.50,0,m,1988,2021-03-01
3,0_1458,2021-03-01 00:04:54.559692,s_4,c_7912,6.55,0,f,1989,2021-03-01
4,0_1358,2021-03-01 00:05:18.801198,s_5,c_2033,16.49,0,f,1956,2021-03-01


In [55]:
#Grouper par mois et sommer les prix dans un nouveau dataframe df_month_price
df_month_price = df.groupby('month').agg({'price': 'sum'}).reset_index()
df_month_price.head()

,month,price
0,2021-03-01,482440.61
1,2021-04-01,476109.30
2,2021-05-01,492943.47
3,2021-06-01,484088.56
4,2021-07-01,482835.40


In [56]:
#Création de la moyenne mobile sur 3 mois
df_month_price['rolling_avg'] = df_month_price['price'].rolling(window=3).mean()

In [ ]:
#Afficher un graphique sur l'évolution dans le temps du chiffre d'affaires avec la moyenne mobile sur 3 mois
fig = px.histogram(df, x='month', y='price', 
              title='Évolution du chiffre d\'affaires par mois avec la moyenne mobile', 
              labels={'month': 'Mois', 'price': 'Chiffre d\'affaires'})
fig.update_layout(xaxis_title='Mois', yaxis_title='Chiffre d\'affaires',
                  title_x=0.5, title_y=0.95)
#Ajouter la moyenne mobile sur le graphique
fig.add_scatter(x=df_month_price['month'], y=df_month_price['rolling_avg'], mode='lines', name='MM', line=dict(color='red', width=2))
fig.update_yaxes(range=[0, df_month_price['price'].max()])
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1.2 - Chiffre d'affaires par catégorie</h4>
</div>

In [ ]:
#Afficher un graphique sur l'évolution dans le temps du chiffre d'affaires par catégorie
fig = px.histogram(df, x='month', y='price', color='categ', 
                   title='Évolution du chiffre d\'affaires par catégorie par mois', 
                   labels={'month': 'Mois', 'price': 'Chiffre d\'affaires', 'categ': 'Catégorie'},
                   color_discrete_map={'0': '#EF553B', '1': '#636EFA', '2': '#00CC96'})
fig.update_layout(xaxis_title='Mois', yaxis_title='Chiffre d\'affaires',
                  title_x=0.5, title_y=0.95)
fig.add_scatter(x=df_month_price['month'], y=df_month_price['rolling_avg'], mode='lines', name='MM', line=dict(color='red', width=2))
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1.3 - Nombre de clients par mois</h4>
</div>

In [57]:
#Grouper par mois et compter les clients dans un nouveau dataframe df_month_clients
df_month_clients = df.groupby('month').agg({'client_id': 'nunique'}).reset_index()
df_month_clients.head()

,month,client_id
0,2021-03-01,5676
1,2021-04-01,5674
2,2021-05-01,5644
3,2021-06-01,5659
4,2021-07-01,5672


In [ ]:
#Afficher un graphique sur l'évolution dans le temps du nombre de clients par mois
fig = px.line(df_month_clients, x='month', y='client_id', 
                   title='Évolution du nombre de clients par mois', 
                   labels={'month': 'Mois', 'client_id': 'Nombre de clients'})
fig.update_layout(xaxis_title='Mois', yaxis_title='Nombre de clients',
                  title_x=0.5, title_y=0.95)
fig.update_yaxes(range=[0, df_month_clients['client_id'].max()])
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1.4 - Nombre de transactions</h4>
</div>

In [58]:
#Grouper par mois et compter les transactions dans un nouveau dataframe df_month_transactions
df_month_transactions = df.groupby('month').agg({'session_id': 'nunique'}).reset_index()
df_month_transactions.head()

,month,session_id
0,2021-03-01,14201
1,2021-04-01,13970
2,2021-05-01,14168
3,2021-06-01,13838
4,2021-07-01,13571


In [ ]:
#Afficher un graphique sur l'évolution dans le temps du nombre de transactions par mois
fig = px.line(df_month_transactions, x='month', y='session_id', 
                   title='Évolution du nombre de transactions par mois', 
                   labels={'month': 'Mois', 'session_id': 'Nombre de transactions'})
fig.update_layout(xaxis_title='Mois', yaxis_title='Nombre de transactions',
                  title_x=0.5, title_y=0.95)
fig.update_yaxes(range=[0, df_month_transactions['session_id'].max()])
fig.show()


<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.1.5 - Nombre de produits vendus</h4>
</div>

In [59]:
#Grouper par mois et compter le nombre de produits vendus dans un nouveau dataframe df_month_products
df_month_products = df.groupby('month').agg({'id_prod': 'count'}).reset_index()
df_month_products.head()

,month,id_prod
0,2021-03-01,28601
1,2021-04-01,28443
2,2021-05-01,28285
3,2021-06-01,26850
4,2021-07-01,24738


In [ ]:
#Afficher un graphique sur l'évolution dans le temps du nombre de transactions par mois
fig = px.line(df_month_products, x='month', y='id_prod', 
                   title='Évolution du nombre de produits vendus par mois', 
                   labels={'month': 'Mois', 'id_prod': 'Nombre de produits vendus'})
fig.update_layout(xaxis_title='Mois', yaxis_title='Nombre de produits vendus',
                  title_x=0.5, title_y=0.95)
fig.update_yaxes(range=[0, df_month_products['id_prod'].max()])
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.2 - Zoom sur les références</h3>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.2.1 - Top des références</h4>
</div>

In [60]:
#Création d'un dataframe top_products avec les produits les plus vendus

'''
#Compter le nombre de ventes de chaque produit
top_products = df['id_prod'].value_counts().reset_index()
#Trier les produits par nombre de ventes
top_products = top_products.sort_values(by='count', ascending=False)
#Ajouter la catégorie des produits
top_products = top_products.merge(products[['id_prod', 'categ']], on='id_prod', how='inner')
top_products.head(10)
'''
#Compter le chiffre d'affaires de chaque produit
top_products = df.groupby('id_prod').agg({'price': 'sum'}).reset_index()
#Trier les produits par chiffre d'affaires
top_products = top_products.sort_values(by='price', ascending=False)
#Ajouter la catégorie des produits
top_products = top_products.merge(products[['id_prod', 'categ']], on='id_prod', how='inner')
top_products.head(10)

,id_prod,price,categ
0,2_159,94893.50,2
1,2_135,69334.95,2
2,2_112,65407.76,2
3,2_102,60736.78,2
4,2_209,56971.86,2
5,1_395,56617.47,1
6,1_369,56136.60,1
7,2_110,53846.25,2
8,1_383,53834.43,1
9,1_414,53522.18,1


In [ ]:
#Afficher un graphique avec le top des 100 produits les plus vendus
fig = px.bar(top_products.head(100), x='id_prod', y='price', color='categ', 
             title='Top 100 des produits les plus vendus', 
             labels={'id_prod': 'ID Produit', 'price': 'Chiffre d\'affaires', 'categ': 'Catégorie'}, 
             color_discrete_map={'0': '#EF553B', '1': '#636EFA', '2': '#00CC96'},
             category_orders={'categ': ['0', '1', '2']})
fig.update_layout(xaxis_title='ID Produit', yaxis_title='Chiffre d\'affaires',
                  title_x=0.5, title_y=0.95)
#Trier les produits par chiffre d'affaires
fig.update_xaxes(categoryorder='total descending')
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.2.2 - Flop des références</h4>
</div>

In [61]:
#Création d'un dataframe top_products avec les produits les moins vendus

'''
#Compter le nombre de ventes de chaque produit
flop_products = df['id_prod'].value_counts().reset_index()
#Trier les produits par nombre de ventes
flop_products = flop_products.sort_values(by='count', ascending=True)
#Ajouter la catégorie des produits
flop_products = flop_products.merge(products[['id_prod', 'categ']], on='id_prod', how='inner')
flop_products.head(20)
'''

#Compter le chiffre d'affaires de chaque produit
flop_products = df.groupby('id_prod').agg({'price': 'sum'}).reset_index()
#Trier les produits par chiffre d'affaires
flop_products = flop_products.sort_values(by='price', ascending=True)
#Ajouter la catégorie des produits
flop_products = flop_products.merge(products[['id_prod', 'categ']], on='id_prod', how='inner')
flop_products.head(10)

,id_prod,price,categ
0,0_1539,0.99,0
1,0_1284,1.38,0
2,0_1653,1.98,0
3,0_1601,1.99,0
4,0_807,1.99,0
5,0_541,1.99,0
6,0_1728,2.27,0
7,0_1498,2.48,0
8,0_898,2.54,0
9,0_1840,2.56,0


In [ ]:
#Afficher un graphique avec le flop des 10 produits les moins vendus
fig = px.bar(flop_products.head(100), x='id_prod', y='price', color='categ', 
             title='Flop 100 des produits les moins vendus', 
             labels={'id_prod': 'ID Produit', 'count': 'Chiffre d\'affaires', 'categ': 'Catégorie'},
             color_discrete_map={'0': '#EF553B', '1': '#636EFA', '2': '#00CC96'}, 
             category_orders={'categ': ['0', '1', '2']})
fig.update_layout(xaxis_title='ID Produit', yaxis_title='Chiffre d\'affaires',
                  title_x=0.5, title_y=0.95)
#Trier les produits par nombre de ventes
fig.update_xaxes(categoryorder='total ascending')
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.2.3 - Répartition par catégorie</h4>
</div>

In [62]:
df['categ'] = df['categ'].astype(str)

In [63]:
sum_products = df.groupby('id_prod').agg({'price': 'sum'}).reset_index()
sum_products = sum_products.merge(products[['id_prod', 'categ']], on='id_prod', how='inner')
sum_products.head()

,id_prod,price,categ
0,0_0,4657.50,0
1,0_1,5352.13,0
2,0_10,394.90,0
3,0_100,61.80,0
4,0_1000,2954.88,0


In [ ]:
#Afficher un graphique sur la répartition des ventes par catégorie de produits
fig = px.pie(sum_products, names='categ', color='categ', values='price',
             title='Répartition du chiffre d\'affaires par catégorie de produits',
             labels={'categ': 'Catégorie','price': 'Chiffre d\'affaires'},
             color_discrete_map={'0': '#EF553B', '1': '#636EFA', '2': '#00CC96'})
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.update_layout(title_x=0.5, title_y=0.95)
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.3 - Profils de clients</h3>
</div>

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.3.1 - Répartition du chiffre d'affaires pour les clients BtoB</h4>
</div>

In [64]:
'''
#Méthode en identifiant les clients B2B de manière arbitraire

#Nous allons d'abord créer des indicateurs qui nous permettront de repérer les clients B2B

#Compter le nombre de transactions par client
nb_transactions = df.groupby('client_id').size().rename('nb_transactions').reset_index()

#Compter le nombre de mois actifs par client
nb_months_active = df.groupby('client_id')['month'].nunique().rename('nb_months_active').reset_index()

#Calculer le montant du panier moyen par client
panier_moyen = df.groupby('client_id')['price'].mean().rename('panier_moyen').reset_index()

#Calculer le volume moyen par produit par client
volume_moyen = df.groupby(['client_id','id_prod']).size().groupby('client_id').mean().rename('volume_moyen').reset_index()

#Rassembler les indicateurs dans un tableau
indicators = pd.concat([nb_transactions, nb_months_active, panier_moyen, volume_moyen], axis=1)

#Définir une fonction pour identifier les clients B2B
def is_b2b(row):
    score = 0
    score += row['nb_transactions'] > 30  # Plus de 30 transactions
    score += row['nb_months_active'] >= 3  # 3 mois actifs ou plus
    score += row['panier_moyen'] > 100  # Panier moyen supérieur à 100
    score += row['volume_moyen'] >= 3  # Volume moyen supérieur à 3
    return True if score >= 3 else False  # Au moins 3 critères remplis

#Appliquer la fonction pour créer une nouvelle colonne 'is_b2b' dans le dataframe customers
customers['is_b2b'] = indicators.apply(is_b2b, axis=1)
customers.loc[customers['is_b2b'].isnull(),'is_b2b'] = False  # Remplacer les valeurs manquantes par False

#Ajouter la colonne 'is_b2b' au dataframe df
df = df.merge(customers[['client_id', 'is_b2b']], on='client_id', how='left')
df.loc[df['is_b2b'].isnull(),'is_b2b'] = False # Remplacer les valeurs manquantes par False
df.head()
'''

"\n#Méthode en identifiant les clients B2B de manière arbitraire\n\n#Nous allons d'abord créer des indicateurs qui nous permettront de repérer les clients B2B\n\n#Compter le nombre de transactions par client\nnb_transactions = df.groupby('client_id').size().rename('nb_transactions').reset_index()\n\n#Compter le nombre de mois actifs par client\nnb_months_active = df.groupby('client_id')['month'].nunique().rename('nb_months_active').reset_index()\n\n#Calculer le montant du panier moyen par client\npanier_moyen = df.groupby('client_id')['price'].mean().rename('panier_moyen').reset_index()\n\n#Calculer le volume moyen par produit par client\nvolume_moyen = df.groupby(['client_id','id_prod']).size().groupby('client_id').mean().rename('volume_moyen').reset_index()\n\n#Rassembler les indicateurs dans un tableau\nindicators = pd.concat([nb_transactions, nb_months_active, panier_moyen, volume_moyen], axis=1)\n\n#Définir une fonction pour identifier les clients B2B\ndef is_b2b(row):\n    score 

In [65]:
#On va essayer d'identifier les clients B2B en fonction de leur chiffre d'affaires
#Grouper par client et sommer les prix dans un nouveau dataframe df_clients
df_clients = df.groupby('client_id').agg({'price': 'sum'}).reset_index()
df_clients.sort_values(by='price', ascending=False, inplace=True)
df_clients.head()

,client_id,price
677,c_1609,326039.89
4388,c_4958,290227.03
6337,c_6714,153918.60
2724,c_3454,114110.57
634,c_1570,5285.82


In [66]:
#On remarque 4 clients atypiques
b2b_clients = ['c_1609','c_4958','c_6714','c_3454']

In [67]:
#On crée une nouvelle colonne 'is_b2b' dans le dataframe df
df['is_b2b'] = df['client_id'].apply(lambda x: 'BtoB' if x in b2b_clients else 'Non BtoB')

In [68]:
df.head()

,id_prod,date,session_id,client_id,price,categ,sex,birth,month,is_b2b
0,0_1259,2021-03-01 00:01:07.843138,s_1,c_329,11.99,0,f,1967,2021-03-01,Non BtoB
1,0_1390,2021-03-01 00:02:26.047414,s_2,c_664,19.37,0,m,1960,2021-03-01,Non BtoB
2,0_1352,2021-03-01 00:02:38.311413,s_3,c_580,4.50,0,m,1988,2021-03-01,Non BtoB
3,0_1458,2021-03-01 00:04:54.559692,s_4,c_7912,6.55,0,f,1989,2021-03-01,Non BtoB
4,0_1358,2021-03-01 00:05:18.801198,s_5,c_2033,16.49,0,f,1956,2021-03-01,Non BtoB


In [ ]:
#Afficher un graphique sur la répartition du chiffre d'affaires pour les clients btob
fig = px.pie(df, names='is_b2b', values='price', color='is_b2b',
             title='Répartition du chiffre d\'affaires pour les clients BtoB',
             labels={'is_b2b': 'Client BtoB', 'price': 'Chiffre d\'affaires (€)'},
             color_discrete_map={'BtoB': '#FFD700', 'Non BtoB': "#9C00CC"})
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.update_layout(legend_title_text='Type de client')
fig.update_layout(title_x=0.5, title_y=0.95)
fig.show()

<div style="border: 1px solid RGB(0,150,250);" >
<h4 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">3.3.2 - Courbe de Lorenz</h4>
</div>

In [69]:
#On trie les clients par chiffre d'affaires
df_clients = df_clients.sort_values(by='price')
df_clients.head()

,client_id,price
8151,c_8351,6.31
7918,c_8140,8.30
7889,c_8114,9.98
4044,c_4648,11.20
3855,c_4478,13.36


In [70]:
#On va calculer les proportions cumulées
df_clients['cumulative'] = df_clients['price'].cumsum() / df_clients['price'].sum()
df_clients['client_pct'] = np.arange(1, len(df_clients) + 1) / len(df_clients)

In [71]:
#On crée df_lorenz en ajouter le point (0,0)
df_lorenz = pd.concat([pd.DataFrame({'client_pct': [0], 'cumulative': [0]}), df_clients[['client_pct', 'cumulative']]], ignore_index=True)

In [72]:
#On calcule l'indice de Gini
gini_index = 1 - 2 * np.trapezoid(df_lorenz['cumulative'], df_lorenz['client_pct'])
print(f"Indice de Gini : {gini_index:.4f}")

Indice de Gini : 0.4419


In [ ]:
#On trace la courbe de Lorenz
fig = px.line(df_lorenz, x='client_pct', y='cumulative',
              title='Courbe de Lorenz du chiffre d\'affaires par client',
              labels={'client_pct': 'Proportion cumulée des clients', 'cumulative': 'Proportion cumulée du chiffre d\'affaires'})
fig.update_layout(title_x=0.5, title_y=0.95)

#On ajoute la ligne d'égalité parfaite
fig.add_scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(dash='dash'), name='Égalité parfaite')

#On ajoute l'indice de Gini
fig.add_annotation(text=f"Indice de Gini : {gini_index:.4f}", xref='paper', yref='paper', x=0.5, y=1.1, showarrow=False)

#On ajoute les points sur la courbe de Lorenz qui passe en x=0.5 ou y=0.5
fig.add_scatter(x=[0.5], y=[df_lorenz[df_lorenz['client_pct'] >= 0.5]['cumulative'].iloc[0]], mode='markers', name='50% des clients', marker=dict(color='yellow', size=10))
fig.add_scatter(x=[df_lorenz[df_lorenz['cumulative'] >= 0.5]['client_pct'].iloc[0]], y=[0.5], mode='markers', name='50% du CA', marker=dict(color='lightgreen', size=10))

fig.show()

<div style="background-color: RGB(0,150,250);" >
<h2 style="margin: auto; padding: 20px; color:#fff; ">4 - Analyse ciblées sur les clients - Corrélations</h2>
</div>

In [73]:
#On crée un dataframe sans les clients BtoB
df_non_b2b = df[df['is_b2b'] == 'Non BtoB'].copy()

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">4.1 - Lien entre le genre d’un client et les catégories des livres achetés</h3>
</div>

In [74]:
from scipy.stats import chi2_contingency

H0 : La catégorie d'un livre acheté est indépendante du genre du client

H1 : La catégorie d'un livre acheté dépend du genre du client 

In [ ]:
# On visualise la relation entre le genre d'un client et les catégories des livres achetés avec une heatmap
fig = px.density_heatmap(
    df_non_b2b,
    x='sex',
    y='categ',
    histfunc='count',
    title="Répartition des catégories de livres achetés par genre (hors BtoB)",
    color_continuous_scale='Blues'
)
#Changer le nom du count
fig.update_layout(xaxis_title='Genre du client', yaxis_title='Catégorie du livre')
fig.update_layout(title_x=0.5, title_y=0.95)
fig.update_layout(coloraxis_colorbar_title='Nombre de livres achetés')
fig.update_traces(texttemplate='%{z}', textfont_size=12)
fig.update_layout(width=600)
fig.show()

In [75]:
#On crée un tableau de contingence pour le genre d'un client et les catégories des livres achetés
contingency_table = pd.crosstab(df_non_b2b['sex'],df_non_b2b['categ'])
print("Tableau de contingence :\n", contingency_table)

Tableau de contingence :
 categ       0       1      2
sex                         
f      200793  115721  16980
m      186488  104884  15868


In [76]:
#On effectue le test du Chi2
chi2, p, dof, expected = chi2_contingency(contingency_table)

#On affiche les résultats du test
print(f"Chi2: {chi2}, p-value: {p}, dof: {dof}")
print("Tableau des fréquences attendues:")
print(expected)

Chi2: 22.66856665178056, p-value: 1.1955928116587024e-05, dof: 2
Tableau des fréquences attendues:
[[201574.89662481 114822.13191434  17096.97146086]
 [185706.10337519 105782.86808566  15751.02853914]]


Hors clients BtoB, on observe une valeur du Chi-2 de 22.7, ce qui représente l'écart entre la table de contingence étudiée et la table des fréquences attendues si les deux variables sont indépendantes.

On observe un degré de liberté de 2, qui suppose que le test est relativement fiable.

Enfin, la p-value de 1.20*10^-5, inférieure au seuil de 0.05, nous confirme qu'on peut rejeter H0, donc qu'il y a une association significative entre le genre d'un client et la catégorie des livres achetés.

In [77]:
contingency_table = pd.crosstab(df['sex'],df['categ'])
print("Tableau de contingence :\n", contingency_table)

Tableau de contingence :
 categ       0       1      2
sex                         
f      206103  119307  17283
m      209356  116285  19200


In [78]:
chi2, p, dof, expected = chi2_contingency(contingency_table)

#On affiche les résultats du test
print(f"Chi2: {chi2}, p-value: {p}, dof: {dof}")
print("Tableau des fréquences attendues:")
print(expected)

Chi2: 158.25417617304882, p-value: 4.3205822283997063e-35, dof: 2
Tableau des fréquences attendues:
[[207080.50959953 117427.98066132  18184.50973915]
 [208378.49040047 118164.01933868  18298.49026085]]


Conclusion similaire en incluant les clients BtoB.

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">4.2 - Lien entre l'âge des clients et le montant total des achats</h3>
</div>

In [79]:
#On crée un dataframe on regroupant les clients par id_client et on garde l'âge et la somme des prix
df_age_price_non_b2b = df_non_b2b.groupby('client_id').agg({'birth': 'first', 'price': 'sum'}).reset_index()
#On crée une nouvelle colonne 'age' en soustrayant l'année de naissance de l'année actuelle
df_age_price_non_b2b['age'] = dt.datetime.now().year - df_age_price_non_b2b['birth']

In [80]:
df_age_price_non_b2b.head()

,client_id,birth,price,age
0,c_1,1955,629.02,71
1,c_10,1956,1353.60,70
2,c_100,1992,254.85,34
3,c_1000,1966,2291.88,60
4,c_1001,1982,1823.85,44


In [81]:
df_age_price_non_b2b.dtypes

client_id     object
birth          int64
price        float64
age            int64
dtype: object

In [82]:
from scipy.stats import normaltest
from scipy.stats import norm

In [83]:
#Appliquer le log sur age
df_age_price_non_b2b['log_age'] = np.log(df_age_price_non_b2b['age'] + 1)  # Ajout de 1 pour éviter log(0)

In [ ]:
#Afficher un graphique sur la distribution de l'âge des clients
fig = px.histogram(df_age_price_non_b2b, x='log_age', nbins=50, 
                   title='Distribution de l\'âge des clients (hors BtoB)', 
                   labels={'log_age': 'Âge', 'count': 'Nombre de clients'})
fig.update_layout(xaxis_title='Âge', yaxis_title='Nombre de clients',
                  title_x=0.5, title_y=0.95)

# Calcul des paramètres de la loi normale
mu, sigma = df_age_price_non_b2b['log_age'].mean(), df_age_price_non_b2b['log_age'].std()
x = np.linspace(df_age_price_non_b2b['log_age'].min(), df_age_price_non_b2b['log_age'].max(), 200)
# Calcul de la densité pour chaque valeur de x
y = norm.pdf(x, mu, sigma)
# Mise à l'échelle pour correspondre à l'histogramme
y_scaled = y * len(df_age_price_non_b2b['log_age']) * (df_age_price_non_b2b['log_age'].max() - df_age_price_non_b2b['log_age'].min()) / 30

fig.add_scatter(x=x, y=y_scaled, mode='lines', name='Courbe normale ajustée', line=dict(color='red'))
fig.show()

In [84]:
stat, p_value = normaltest(df_age_price_non_b2b['log_age'])

print(f"Statistique du test de normalité : {stat}")
print(f"Valeur p : {p_value}")

Statistique du test de normalité : 1042.445099846167
Valeur p : 4.324368795797162e-227


In [85]:
#Appliquer le log sur price
df_age_price_non_b2b['log_price'] = np.log(df_age_price_non_b2b['price'] + 1)  # Ajout de 1 pour éviter log(0)

In [ ]:
fig = px.histogram(df_age_price_non_b2b, x='log_price', nbins=50, 
                   title='Distribution du montant total du panier (hors BtoB)', 
                   labels={'log_price': 'Montant', 'count': 'Nombre de clients'})
fig.update_layout(xaxis_title='Montant total du panier', yaxis_title='Nombre de clients',
                  title_x=0.5, title_y=0.95)

mu, sigma = df_age_price_non_b2b['log_price'].mean(), df_age_price_non_b2b['log_price'].std()
x = np.linspace(df_age_price_non_b2b['log_price'].min(), df_age_price_non_b2b['log_price'].max(), 200)
y = norm.pdf(x, mu, sigma)
y_scaled = y * len(df_age_price_non_b2b['log_price']) * (df_age_price_non_b2b['log_price'].max() - df_age_price_non_b2b['log_price'].min()) / 30

fig.add_scatter(x=x, y=y_scaled, mode='lines', name='Courbe normale ajustée', line=dict(color='red'))
fig.show()

In [86]:
stat, p_value = normaltest(df_age_price_non_b2b['log_price'])

print(f"Statistique du test de normalité : {stat}")
print(f"Valeur p : {p_value}")

Statistique du test de normalité : 810.2135801146904
Valeur p : 1.1597289448801489e-176


In [87]:
from scipy.stats import pearsonr

H0 = Il n’y a pas de corrélation linéaire entre l'âge des clients et le montant total des achats.

H1 = Il y a une corrélation linéaire entre l'âge des clients et le montant total des achats.

In [ ]:
#Nuage de points entre l'âge des clients et le montant total des achats avec une régression linéaire
fig = px.scatter(
    df_age_price_non_b2b,
    x='age',
    y='price',
    trendline='ols',
    title='Nuage de points entre l\'âge des clients et le montant total des achats',
    labels={'age': 'Âge du client', 'price': 'Montant total des achats (€)'}
)
fig.update_layout(title_x=0.5, title_y=0.95)
fig.update_traces(line=dict(color='red'))
fig.update_layout(width=800)
fig.show()


In [88]:
#On effectue le test de Pearson
correlation, p_value = pearsonr(df_age_price_non_b2b["age"], df_age_price_non_b2b["price"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: -0.18756654332685188
Valeur p: 6.66351155977455e-69


Hors clients Btob, on observe un coefficient de -0.19, ce qui indique une faible corrélation négative entre l'âge du client et le montant total d'achat, plus le client est âgé, plus le montant total des achats est bas.

La p-value très faible de 6.67*10^-69, inférieure à 0.05, nous permet de rejeter H0, donc il y a bien une corrélation linéaire entre l'âge du client et le montant total d'achat.

In [89]:
df_age_price = df.groupby('client_id').agg({'birth': 'first', 'price': 'sum'}).reset_index()
df_age_price['age'] = dt.datetime.now().year - df_age_price['birth']

correlation, p_value = pearsonr(df_age_price["age"], df_age_price["price"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: -0.03975597869436891
Valeur p: 0.0002262549055777749


Conclusion similaire en incluant les clients BtoB.

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">4.3 - Lien entre l'âge des clients et la fréquence d’achat</h3>
</div>

In [90]:
#On crée un dataframe on regroupant les clients par id_client et on garde l'âge et la fréquence d'achat
df_age_freq_non_b2b = df_non_b2b.groupby('client_id').agg({'birth': 'first', 'session_id': 'nunique'}).reset_index()
#On crée une nouvelle colonne 'age' en soustrayant l'année de naissance de l'année actuelle
df_age_freq_non_b2b['age'] = dt.datetime.now().year - df_age_freq_non_b2b['birth']

In [91]:
df_age_freq_non_b2b.head()

,client_id,birth,session_id,age
0,c_1,1955,34,71
1,c_10,1956,34,70
2,c_100,1992,5,34
3,c_1000,1966,94,60
4,c_1001,1982,47,44


In [92]:
df_age_freq_non_b2b.dtypes

client_id     object
birth          int64
session_id     int64
age            int64
dtype: object

In [93]:
#Appliquer le log sur session_id
df_age_freq_non_b2b['log_session_id'] = np.log(df_age_freq_non_b2b['session_id'] + 1)  # Ajout de 1 pour éviter log(0)

In [ ]:
fig = px.histogram(df_age_freq_non_b2b, x='log_session_id', nbins=30, 
                   title='Distribution de la fréquence d\'achat (hors BtoB)', 
                   labels={'log_session_id': 'Fréquence d\'achat', 'count': 'Nombre de clients'})
fig.update_layout(xaxis_title='Fréquence d\'achat', yaxis_title='Nombre de clients',
                  title_x=0.5, title_y=0.95)

mu, sigma = df_age_freq_non_b2b['log_session_id'].mean(), df_age_freq_non_b2b['log_session_id'].std()
x = np.linspace(df_age_freq_non_b2b['log_session_id'].min(), df_age_freq_non_b2b['log_session_id'].max(), 200)
y = norm.pdf(x, mu, sigma)
y_scaled = y * len(df_age_freq_non_b2b['log_session_id']) * (df_age_freq_non_b2b['log_session_id'].max() - df_age_freq_non_b2b['log_session_id'].min()) / 30

fig.add_scatter(x=x, y=y_scaled, mode='lines', name='Courbe normale ajustée', line=dict(color='red'))
fig.show()

In [94]:
stat, p_value = normaltest(df_age_freq_non_b2b['log_session_id'])

print(f"Statistique du test de normalité : {stat}")
print(f"Valeur p : {p_value}")

Statistique du test de normalité : 114.0556857209188
Valeur p : 1.7104978429859686e-25


H0 = Il n’y a pas de corrélation linéaire entre l'âge des clients et la fréquence d'achat.

H1 = Il y a une corrélation linéaire entre l'âge des clients et la fréquence d'achat.

In [ ]:
#On crée un nuage de points entre l'âge des clients et la fréquence d'achat avec une régression linéaire rouge
fig = px.scatter(
    df_age_freq_non_b2b,
    x='age',
    y='session_id',
    trendline='ols',
    title='Nuage de points entre l\'âge des clients et la fréquence d\'achat',
    labels={'age': 'Âge du client', 'session_id': 'Fréquence d\'achat'}
)
fig.update_layout(title_x=0.5, title_y=0.95)
fig.update_traces(line=dict(color='red'))
fig.update_layout(width=800)
fig.show()

In [95]:
#On effectue le test de Pearson
correlation, p_value = pearsonr(df_age_freq_non_b2b["age"], df_age_freq_non_b2b["session_id"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: 0.16449131101203116
Valeur p: 3.3704380510771483e-53


Hors clients Btob, on observe un coefficient de 0.16, ce qui indique une faible corrélation positive entre l'âge du client et la fréquence d'achat, plus le client est âgé, plus souvent il achètera.

La p-value très faible de 3.37*10^-53, inférieure à 0.05, nous permet de rejeter H0, donc il y a bien une corrélation linéaire entre l'âge du client et le montant total d'achat.

In [96]:
df_age_freq = df.groupby('client_id').agg({'birth': 'first', 'session_id': 'nunique'}).reset_index()
df_age_freq['age'] = dt.datetime.now().year - df_age_freq['birth']

correlation, p_value = pearsonr(df_age_freq["age"], df_age_freq["session_id"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: 0.03413851974617009
Valeur p: 0.001543709016583611


Conclusion similaire en incluant les clients BtoB.

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">4.4 - Lien entre l'âge des clients et la taille du panier moyen</h3>
</div>

In [97]:
#On crée un dataframe on regroupant les clients par id_client et on garde l'âge et la taille du panier moyen
df_age_panier_non_b2b = df_non_b2b.groupby('client_id').agg({'birth': 'first', 'price': 'mean'}).reset_index()
#On crée une nouvelle colonne 'age' en soustrayant l'année de naissance de l'année actuelle
df_age_panier_non_b2b['age'] = dt.datetime.now().year - df_age_panier_non_b2b['birth']

In [98]:
df_age_panier_non_b2b.head()

,client_id,birth,price,age
0,c_1,1955,14.628372,71
1,c_10,1956,23.337931,70
2,c_100,1992,31.856250,34
3,c_1000,1966,18.189524,60
4,c_1001,1982,17.707282,44


In [99]:
df_age_panier_non_b2b.dtypes

client_id     object
birth          int64
price        float64
age            int64
dtype: object

In [100]:
#Appliquer le log sur price
df_age_panier_non_b2b['log_price'] = np.log(df_age_panier_non_b2b['price'] + 1)  # Ajout de 1 pour éviter log(0)

In [ ]:
fig = px.histogram(df_age_panier_non_b2b, x='log_price', nbins=100, 
                   title='Distribution de la taille du panier moyen (hors BtoB)', 
                   labels={'log_price': 'Taille du panier moyen', 'count': 'Nombre de clients'})
fig.update_layout(xaxis_title='Taille du panier moyen', yaxis_title='Nombre de clients',
                  title_x=0.5, title_y=0.95)

mu, sigma = df_age_panier_non_b2b['log_price'].mean(), df_age_panier_non_b2b['log_price'].std()
x = np.linspace(df_age_panier_non_b2b['log_price'].min(), df_age_panier_non_b2b['log_price'].max(), 200)
y = norm.pdf(x, mu, sigma)
y_scaled = y * len(df_age_panier_non_b2b['log_price']) * (df_age_panier_non_b2b['log_price'].max() - df_age_panier_non_b2b['log_price'].min()) / 30

fig.add_scatter(x=x, y=y_scaled, mode='lines', name='Courbe normale ajustée', line=dict(color='red'))
fig.show()

In [101]:
stat, p_value = normaltest(df_age_panier_non_b2b['price'])

print(f"Statistique du test de normalité : {stat}")
print(f"Valeur p : {p_value}")

Statistique du test de normalité : 4558.283005580248
Valeur p : 0.0


H0 = Il n’y a pas de corrélation linéaire entre l'âge des clients et la taille du panier moyen.

H1 = Il y a une corrélation linéaire entre l'âge des clients et la taille du panier moyen.

In [ ]:
#On crée un nuage de points entre l'âge des clients et la taille du panier moyen avec une régression linéaire rouge
fig = px.scatter(
    df_age_panier_non_b2b,
    x='age',
    y='price',
    trendline='ols',
    title='Nuage de points entre l\'âge des clients et la taille du panier moyen',
    labels={'age': 'Âge du client', 'price': 'Taille du panier moyen (€)'}
)
fig.update_layout(title_x=0.5, title_y=0.95)
fig.update_traces(line=dict(color='red'))
fig.update_layout(width=800)
fig.show()

In [102]:
#On effectue le test de Pearson
correlation, p_value = pearsonr(df_age_panier_non_b2b["age"], df_age_panier_non_b2b["price"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: -0.5105290328181461
Valeur p: 0.0


Hors clients Btob, on observe un coefficient de -0.51, ce qui indique une corrélation négative entre l'âge du client et la taille du panier moyen, plus le client est âgé, plus son panier moyen sera bas.

La p-value nulle, inférieure à 0.05, nous permet de rejeter H0, donc il y a bien une corrélation linéaire entre l'âge du client et le montant total d'achat.

In [103]:
df_age_panier = df.groupby('client_id').agg({'birth': 'first', 'price': 'mean'}).reset_index()
df_age_panier['age'] = dt.datetime.now().year - df_age_panier['birth']

correlation, p_value = pearsonr(df_age_panier["age"], df_age_panier["price"])

print(f"Coefficient de corrélation de Pearson: {correlation}")
print(f"Valeur p: {p_value}")

Coefficient de corrélation de Pearson: -0.5106560259205026
Valeur p: 0.0


Conclusion similaire en incluant les clients BtoB.

<div style="border: 1px solid RGB(0,150,250);" >
<h3 style="margin: auto; padding: 20px; color: RGB(0,150,250); ">4.5 - Lien entre l'âge des clients et la catégorie des livres achetés</h3>
</div>

In [104]:
from scipy.stats import f_oneway

In [105]:
df_non_b2b['age'] = dt.datetime.now().year - df_non_b2b['birth']
df['age'] = dt.datetime.now().year - df['birth']

H0 : Toutes les moyennes d'âge de chaque catégorie de livres achetés sont égales.

H0 : Au moins une moyenne d'âge est différente.

In [ ]:
#Boxplot de l'âge des clients par catégorie de produits
fig = px.box(df_non_b2b, x='categ', y='age', 
             title='Boxplot de l\'âge des clients par catégorie de produits (hors BtoB)',
             labels={'categ': 'Catégorie de produits', 'age': 'Âge du client'},
             color='categ', color_discrete_map={'0': '#EF553B', '1': '#636EFA', '2': '#00CC96'})
fig.update_layout(xaxis_title='Catégorie de produits', yaxis_title='Âge du client',
                  title_x=0.5, title_y=0.95)
fig.update_traces(boxpoints=False)
fig.update_layout(width=800)
fig.show()

In [106]:
categ_0 = df_non_b2b[df_non_b2b['categ'] == '0']['age']
categ_1 = df_non_b2b[df_non_b2b['categ'] == '1']['age']
categ_2 = df_non_b2b[df_non_b2b['categ'] == '2']['age']

print("Moyenne de l'âge pour la catégorie 0 : ", categ_0.mean())
print("Moyenne de l'âge pour la catégorie 1 : ", categ_1.mean())
print("Moyenne de l'âge pour la catégorie 2 : ", categ_2.mean())

Moyenne de l'âge pour la catégorie 0 :  47.79456260441385
Moyenne de l'âge pour la catégorie 1 :  51.650035130663404
Moyenne de l'âge pour la catégorie 2 :  29.965507793472966


In [107]:
f_stat, p_value = f_oneway(categ_0, categ_1, categ_2)

print(f"Statistique de test F: {f_stat}")
print(f"Valeur p: {p_value}")

Statistique de test F: 39705.51969324288
Valeur p: 0.0


Hors clients Btob, on observe une statistique de test de 39 705.52, ce qui indique qu'il y a une grande différence entre les groupes.

La p-value nulle, inférieure au seuil de 0.05, nous permet de rejeter H0, donc il y a bien une différence significative d'âge moyen entre au moins deux catégories de livres achetés.

In [108]:
categ_0 = df[df['categ'] == '0']['age']
categ_1 = df[df['categ'] == '1']['age']
categ_2 = df[df['categ'] == '2']['age']

print("Moyenne de l'âge pour la catégorie 0 : ", categ_0.mean())
print("Moyenne de l'âge pour la catégorie 1 : ", categ_1.mean())
print("Moyenne de l'âge pour la catégorie 2 : ", categ_2.mean())

f_stat, p_value = f_oneway(categ_0, categ_1, categ_2)

print(f"Statistique de test F: {f_stat}")
print(f"Valeur p: {p_value}")

Moyenne de l'âge pour la catégorie 0 :  47.89744595736282
Moyenne de l'âge pour la catégorie 1 :  51.50923206220924
Moyenne de l'âge pour la catégorie 2 :  29.94803058958967
Statistique de test F: 45120.035535673815
Valeur p: 0.0


Conclusion similaire en incluant les clients BtoB.